### ISSUE TITLE
Iso 0°C data seems incorrect

### ISSUE THREAD

**Issue by Oeilducyclone** [open-meteo/open-meteo#518]
Hi !
I noticed that the Iso 0°C data seems incorrect.
It is apparently often too high (3000m when it should be 1500m) and it seems it doesn't get lower than the altitude of the ref point.
IE : For a point with an elevation of 1200m and a temp of -2.5°C the Iso 0 level is 1200m ... So when the freezing level is all day under 1200m ... It shows 1200m all day !
Best regards
Guillaume

**Comment by patrick-zippenfenig** [comment]
For which location did you observe this error? I just tested in in Central Europe and it seems to be fine.

I confirm, that freezing level height does not go below reference height (grid-cell elevation in this case).

**Comment by Oeilducyclone** [comment]
Hi Patrick,
Thank you for your reply.
I get weird datas for many locations.
IE : Val d'Isère (1850m) or Tignes (2050m) ski stations.
For Val d'Isère (1850m) I have Today (Dim Nov 26) a T° of -10°C at 8am and a freezing level at 2160m (All day).
Things are even stranger with Nov28 when freezing level get from 2160m to 2470m at 8am with a temp between -3 and -7°C all day...
Is there any way I can get the real freezing level (With no ref height cap)?

**Comment by patrick-zippenfenig** [comment]
Ah, now I get the issue. Those 3 steps are caused by combining local, continental and global models. The local ICON-D2 model only offers 3 days of forecast, the continental ICON-EU 5 days and the global ICON model up to 8 days of forecast.

Each model uses different grid resolution and topography. Therefore the reference height is 2160, 2470 and 2100 meters. See below. The middle peak is freezing height above reference height.

![Screenshot 2023-11-27 at 09 32 40](https://github.com/open-meteo/open-meteo/assets/11670456/5d0399e1-a999-49cb-9457-5c0257deee08)

Because the target elevation is 1891 m, but each model grid cell is above 2000 meter, there is no information available to calculate the freezing level height correctly. 

One option: The 2 meter above ground temperature could be used to extrapolate the freezing level height below grid-cell elevation. E.g. if the freezing level height is at 2160 m (= grid elevation) at -5°C it can be assumed that the freezing level height is at 1446 m above mean sea level (`2160 m - 5°C * 0.7°C/100 m = 1446 m`). This is ignoring any kind of temperature inversion and assumes a leaps rate of 0.7°C.

I might implement this correction in the next days

**Comment by patrick-zippenfenig** [comment]
> The CFS is providing the T° at different pressure levels 500/700/850/900 ... This may be a way to get more accurate datas in case of inversion.

Weather models nowadays use model and hybrid levels that to simulate weather on atmospheric layers. Pressure level are calculated in post-processing. Because model layers are bound to ground, any pressure levels below ground do not provide reliable data.

I implemented the lapse rate correction in #524. However, it will take a day until results are live.

Preliminary results look promising:

Before:
![Screenshot 2023-11-27 at 12 31 46](https://github.com/open-meteo/open-meteo/assets/11670456/5bcc47df-540b-4824-9f0f-4b849b74d16c)

After:
![Screenshot 2023-11-27 at 12 31 51](https://github.com/open-meteo/open-meteo/assets/11670456/e7d281eb-319d-4e67-a5e0-e964c95aa358)

Note: Data is not exactly the same, as I used different model runs locally
